students/yushus/session05/mailroom3.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env/python3
+
+"""
+Yushu Song
+Au2018-Py210B
+Mailroom3 assignment
+"""
+
+import os
+import random
+import sys
+
+donor_db = {"William Gates, III": [653772.32, 12.17],
+            "Jeff Bezos": [877.33],
+            "Paul Allen": [663.23, 43.87, 1.32],
+            "Mark Zuckerberg": [1663.23, 4300.87, 10432.0],
+            "Elon Musk": [250000, 15000]}
+
+prompt = "\n".join(("Welcome to mailroom bot!",
+          "Please choose from below options:",
+          "1 - Send a Thank You",
+          "2 - Create a Report",
+          "3 - Send letters to all donors",
+          "4 - quit",
+          ">>> "))
+
+def print_donor():
+    print("\n".join(donor_db.keys()))
+
+def format_letter():
+    letter = "\n".join(("Dear {name},\n",
+             "\tThank you for your generous donation of ${amount}.\n",
+             "\tYour kindness is really making the world different!\n",
+             "\t\tSincerely,",
+             "\t\tMailroom Bot"))
+    
+    return letter
+
+def print_email(name, amount):
+    print(format_letter().format(name=name, amount=amount))
+
+# Send a thank you note for the donor
+# The donor name is case insensitive
+# However, the donor_db will keep the same case when donor name was first entered to the system
+def send_a_thank_you():
+
+    # maintain a donor list which contains donor names only
+    donors = [d.lower() for d in donor_db.keys()]
+    name = ""
+
+    while True:
+        name = input("Who do you want to send this to, use a full name: ")
+        if name.lower() == 'list':
+            print_donor()
+        elif name.lower() not in donors: # New donor
+            donor_db[name]=[]
+            break
+        else: # Existing donor; find the correct donor name from DB
+            for donor_name in donor_db:
+                if donor_name.lower() == name.lower():
+                    name = donor_name
+            break
```

Comment:
Hmm, I must have overlooked this part. Once you have a dict for `donor_db`, this `for` loop is not really needed. Remember that the only reason we had this loop was because with a list for `donor_db`, we couldn't say `donor_db[name]`, so we had to go through all items in the list and find out the matching one. With a dict, there's no need to go through all items.

This is related to the earlier assignment, not this one, so this change is not required, but I'd say it'd be good to do that anyway. I'll leave it up to you.

Reply:
The thing I am trying to achieve is to ignore cases when user provides the donor name. Since dict is case sensitive for the keys, I have to find the "original" key by loop through the dict.

Reply:
I see. It's more like a data/input sanitization issue. In that case, I suppose it'd be still better to avoid this kind of sequential search with a dict, by for example, standardizing the keys (like `str.title()`?). If you have to do sequential search like this with a dict, there's really no point using a dict at all.